flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java

```diff
@@ -209,4 +211,15 @@ protected void handleStateChange(ConnectionState newState) {
 	public void unhandledError(String s, Throwable throwable) {
 		leaderListener.handleError(new FlinkException("Unhandled error in ZooKeeperLeaderRetrievalService:" + s, throwable));
 	}
+
+	private void notifyLeaderLoss() {
+		if (lastLeaderAddress != null || lastLeaderSessionID != null) {
+			LOG.debug(
+					"No leader information could be retrieved. Any listeners will be notified.");
+
+			lastLeaderAddress = null;
+			lastLeaderSessionID = null;
```

Comment:
The state changes are not happening under the `lock`. If you introduce `notifyIfNewLeaderAddress`, then this call should happen under the `lock` and we could add `@GuardedBy("lock")` to `notifyIfNewLeaderAddress` to state this contract.